### PR TITLE
PIM-6455: Update family variants through imports

### PR DESCRIPTION
**Pull request by damien-carcel**
## Description

This PR reworks the `FamilyVariantUpdater` so we can update existing family variants, and add new business rules specific to the update:
- the family of the family variant cannot be changed,
- the number of level cannot be changed,
- the axes cannot be changed,
- attributes (if not axes) can be moved to lower attribute set, but not to upper.

Label of the CSV and XLSX import jobs are also updated to be compliant with existing job labels.

## Definition Of Done

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added Behats                      | OK
| Added integration tests           | OK
| Changelog updated                 | -
| Review and 2 GTM                  | OK
| Micro Demo to the PO (Story only) | OK
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Component/Catalog/Model/FamilyVariantInterface.php [hunk 1]
```diff
@@ -44,6 +44,11 @@ public function getVariantAttributeSet(int $level): ?VariantAttributeSetInterfac
     /**
      * @return Collection
      */
+    public function getVariantAttributeSets(): Collection;
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Model/FamilyVariantInterface.php]: I didn't find where you use it
so remove it please :)

File: src/Pim/Component/Catalog/Model/FamilyVariant.php [hunk 2]
```diff
@@ -105,6 +105,14 @@ public function getVariantAttributeSet(int $level): ?VariantAttributeSetInterfac
     /**
      * {@inheritdoc}
      */
+    public function getVariantAttributeSets(): Collection
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Model/FamilyVariant.php]: I didn't find where you use it
so remove it please :)

File: src/Pim/Bundle/CatalogBundle/Resources/config/validation/familyvariant.yml [hunk 3]
```diff
@@ -16,6 +17,10 @@ Pim\Component\Catalog\Model\FamilyVariant:
                 max: 100
         translations:
             - Valid: ~
+    getters:
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Resources/config/validation/familyvariant.yml]: I didn't find where you use it
so remove it please :)

Comment (damien-carcel) [inline src/Pim/Bundle/CatalogBundle/Resources/config/validation/familyvariant.yml]: This is used so the familyvariant validator can also validate in cascade the attribute sets.
And for that, I need the `getVariantAttributeSets` method.

File: src/Pim/Component/Catalog/Model/FamilyVariant.php (hunk 2 above)
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Model/FamilyVariant.php]: Answered in next comment.

File: src/Pim/Bundle/CatalogBundle/tests/integration/Family/UpdateFamilyVariantIntegration.php [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+                    'attributes' => ['EAN', 'brand', 'collection', 'color', 'description', 'erp_name', 'image_1'],
+                    'level'=> 1,
+                ],
+            ],
+        ]);
+
+        $errors = $this->validateFamilyVariant($familyVariant);
+        $this->assertEquals(0, $errors->count());
+
+        $this->get('pim_catalog.saver.family_variant')->save($familyVariant);
```
Comment (aRn0D) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Family/UpdateFamilyVariantIntegration.php]: Julien did not want that an integration test tests saving and loading (you should have a look to the creation one).

File: src/Pim/Bundle/CatalogBundle/tests/integration/Family/UpdateFamilyVariantIntegration.php [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+        $this->assertEquals(
+            'Attributes must be unique, "weight" are used several times in variant attributes sets',
+            $errors->get(0)->getMessage()
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getConfiguration()
```
Comment (aRn0D) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Family/UpdateFamilyVariantIntegration.php]: return typehint

File: src/Pim/Component/Catalog/Model/VariantAttributeSet.php [hunk 6]
```diff
@@ -59,7 +59,11 @@ public function hasAttribute(AttributeInterface $attribute): bool
      */
     public function setAttributes(array $attributes): void
     {
-        $this->attributes = new ArrayCollection($attributes);
+        foreach ($attributes as $attribute) {
```
Comment (aRn0D) [inline src/Pim/Component/Catalog/Model/VariantAttributeSet.php]: Should we create a collection with unique objects?

File: src/Pim/Component/Catalog/Validator/Constraints/FamilyVariantValidator.php [hunk 7]
```diff
@@ -2,6 +2,8 @@
 
 namespace Pim\Component\Catalog\Validator\Constraints;
 
+use Doctrine\ORM\EntityManagerInterface;
```
Comment (aRn0D) [inline src/Pim/Component/Catalog/Validator/Constraints/FamilyVariantValidator.php]: Our `Domain` must not depend on the `Infrastructure`!

File: src/Pim/Component/Catalog/Updater/FamilyVariantUpdater.php [hunk 8]
```diff
@@ -193,13 +197,26 @@ private function removeAttributeFromPreviousLevel(
...
      */
-    private function getAttributes(array $attributeCodes): array
+    private function getAttributes(array $attributeCodes, int $level): array
     {
-        return array_map(function ($attributeCode) {
-            return $this->attributeRepository->findOneByIdentifier($attributeCode);
+        return array_map(function ($attributeCode) use ($level) {
+            $attribute = $this->attributeRepository->findOneByIdentifier($attributeCode);
+
+            if (!$attribute instanceof AttributeInterface) {
```
Comment (aRn0D) [inline src/Pim/Component/Catalog/Updater/FamilyVariantUpdater.php]: null !== $attribute

File: src/Pim/Bundle/CatalogBundle/Resources/config/validators.yml [hunk 9]
```diff
@@ -67,7 +67,6 @@ services:
         class: '%pim_catalog.validator.constraint.family_variant_axes.class%'
         arguments:
             - '@translator'
-            - '@doctrine.orm.entity_manager'
```
Comment (aRn0D) [inline src/Pim/Bundle/CatalogBundle/Resources/config/validators.yml]: :+1:

File: src/Pim/Component/Catalog/Updater/FamilyVariantUpdater.php (hunk 8 above)
Comment (jjanvier) [inline src/Pim/Component/Catalog/Updater/FamilyVariantUpdater.php]: `!$bidule` is for booleans :see_no_evil:

**Pull request merged**